pmda/density.py

```diff
@@ -0,0 +1,222 @@
+        super(pDensity, self).__init__(u, (atomgroup, ))
+        self._atomgroup = atomgroup
+        self._trajectory = u.trajectory
+        self._cutoff = cutoff
+        self._atomselection = atomselection
+        self._soluteselection = soluteselection
+        self._update_selection = update_selection
+        self._padding = padding
+        self._metadata = metadata
+        self._parameters = parameters
+        self._n_frames = u.trajectory.n_frames
+        coord = self.current_coordinates(atomgroup, cutoff, atomselection, soluteselection, update_selection)
+        box, angles = self._trajectory.ts.dimensions[:3], self._trajectory.ts.dimensions[3:]
+        # logger.info(
+        #     "Selected {0:d} atoms out of {1:d} atoms ({2!s}) from {3:d} total."
+        #     "".format(coord.shape[0], len(u.select_atoms(atomselection)),
+        #               atomselection, len(u.atoms))
+        # )
+
+        # mild warning; typically this is run on RMS-fitted trajectories and
+        # so the box information is rather meaningless
+        # if tuple(angles) != (90., 90., 90.):
+        #     msg = "Non-orthorhombic unit-cell --- make sure that it has been remapped properly!"
+        #     warnings.warn(msg)
+        #     logger.warning(msg)
+
+        if gridcenter is not None:
+            # Generate a copy of smin/smax from coords to later check if the
+            # defined box might be too small for the selection
+            smin = np.min(coord, axis=0)
+            smax = np.max(coord, axis=0)
+            # Overwrite smin/smax with user defined values
+            smin, smax = _set_user_grid(gridcenter, xdim, ydim, zdim, smin, smax)
+        else:
+            # Make the box bigger to avoid as much as possible 'outlier'. This
+            # is important if the sites are defined at a high density: in this
+            # case the bulk regions don't have to be close to 1 * n0 but can
+            # be less. It's much more difficult to deal with outliers.  The
+            # ideal solution would use images: implement 'looking across the
+            # periodic boundaries' but that gets complicate when the box
+            # rotates due to RMS fitting.
+            smin = np.min(coord, axis=0) - padding
+            smax = np.max(coord, axis=0) + padding
+            BINS = fixedwidth_bins(delta, smin, smax)
+            arange = np.transpose(np.vstack((BINS['min'], BINS['max'])))
+            bins = BINS['Nbins']
+            # create empty grid with the right dimensions (and get the edges)
+            grid, edges = np.histogramdd(np.zeros((1, 3)), bins=bins, range=arange, normed=False)
+            grid *= 0.0
+        self._grid = grid
+        self._edges = edges
+        h = grid.copy()
+        self._arange = arange
+        self._bins = bins
+
+    def _prepare(self):
+        """
+        Additional preparation to run
+        """
+
+        # pm = ProgressMeter(self._n_frames, interval=interval,
```

Comment:
remove pm

Reply:
I would have expected `_prepare`  to contain all the setup from `__init__`. 

- write test with this working code
- move stuff back into `_prepare`